sklearn/metrics/tests/test_pairwise_distances_reduction.py

```diff
@@ -96,24 +193,527 @@ def assert_radius_neighborhood_results_equality(ref_dist, dist, ref_indices, ind
+                ]
+            ),
+            rtol=rtol,
+        )
+
+    # Indices aren't properly sorted w.r.t their distances
+    msg = "Neighbors indices for query 0 are not matching"
+    with pytest.raises(AssertionError, match=msg):
+        assert_argkmin_results_quasi_equality(
+            ref_dist=np.array(
+                [
+                    [1.2, 2.5, _6_1m, 6.1, _6_1p],
+                ]
+            ),
+            dist=np.array(
+                [
+                    [1.2, 2.5, _6_1m, 6.1, _6_1p],
+                ]
+            ),
+            ref_indices=np.array(
+                [
+                    [1, 2, 3, 4, 5],
+                ]
+            ),
+            indices=np.array(
+                [
+                    [2, 1, 4, 5, 3],
+                ]
+            ),
+            rtol=rtol,
+        )
+
+    # Distances aren't properly sorted
+    msg = "Distances aren't sorted on row 0"
+    with pytest.raises(AssertionError, match=msg):
+        assert_argkmin_results_quasi_equality(
+            ref_dist=np.array(
+                [
+                    [1.2, 2.5, _6_1m, 6.1, _6_1p],
+                ]
+            ),
+            dist=np.array(
+                [
+                    [2.5, 1.2, _6_1m, 6.1, _6_1p],
+                ]
+            ),
+            ref_indices=np.array(
+                [
+                    [1, 2, 3, 4, 5],
+                ]
+            ),
+            indices=np.array(
+                [
+                    [2, 1, 4, 5, 3],
+                ]
+            ),
+            rtol=rtol,
+        )
+
+
+def test_assert_radius_neighborhood_results_quasi_equality():
+
+    rtol = 1e-7
+    atol = 1e-7
+
+    _1m = 1.0 - atol
+    _1p = 1.0 + atol
+
+    _6_1m = 6.1 - atol
+    _6_1p = 6.1 + atol
+
+    ref_dist = np.array(
+        [
+            np.array([1.2, 2.5, _6_1m, 6.1, _6_1p]),
+            np.array([_1m, 1, _1p, _1p]),
+        ]
+    )
+    ref_indices = np.array(
+        [
+            np.array([1, 2, 3, 4, 5]),
+            np.array([6, 7, 8, 9]),
+        ]
+    )
+
+    # Sanity check: compare the reference results to themselves.
+    assert_radius_neighborhood_results_quasi_equality(
+        ref_dist,
+        ref_dist,
+        ref_indices,
+        ref_indices,
+        radius=6.1,
+        rtol=rtol,
+    )
+
+    # Apply valid permutation on indices
+    assert_radius_neighborhood_results_quasi_equality(
+        ref_dist=np.array(
+            [
+                np.array([1.2, 2.5, _6_1m, 6.1, _6_1p]),
+            ]
+        ),
+        dist=np.array(
+            [
+                np.array([1.2, 2.5, _6_1m, 6.1, _6_1p]),
+            ]
+        ),
+        ref_indices=np.array(
+            [
+                np.array([1, 2, 3, 4, 5]),
+            ]
+        ),
+        indices=np.array(
+            [
+                np.array([1, 2, 4, 5, 3]),
+            ]
+        ),
+        radius=6.1,
+        rtol=rtol,
+    )
```

Comment:
It's possible to write the previous assertion more compactly as:

```python
    assert_radius_neighborhood_results_quasi_equality(
        np.array([np.array([1.2, 2.5, _6_1m, 6.1, _6_1p])]),
        np.array([np.array([1.2, 2.5, _6_1m, 6.1, _6_1p])]),
        np.array([np.array([1, 2, 3, 4, 5])]),
        np.array([np.array([1, 2, 4, 5, 3])]),
        radius=6.1,
        rtol=rtol,
    )
```

I removed the trailing `,` in the nested arrays and also removed the keyword specifiers for the positional args.

This is more compact while still benefiting from the vertical alignment of permuted values but it's less explicit as to what are the ref dists and indices. But maybe this is fine.

I don't have strong opinions on which style is most readable.